### DISCUSSION TITLE
Include related document including counters?

### DISCUSSION THREAD

**Discussion by SbiCA** [ravendb/ravendb#15965] [Q&A] [answered]
Hey there, 

Is it possible to have a query such as 

``` c#
   var memberships = await session.Query<MembersBy_EventAndUser.Result, MembersBy_EventAndUser>()
            .Where(mem => mem.UserId == currentUser && !mem.IsDeleted)
            .OfType<MembershipDocument>()
            .Include(mem => mem.EventId) // TODO LEVEL 2 I'd like the counters of this document to be included?? 🤔
            .Include(mem => mem.IncludeAllCounters()) // LEVEL 1 ✅
            .Include(mem => mem.UserProfileId)
            .ToListAsync();
```  

Where we fetch the membership(s) of a user and like to load all the related event documents (which includes counters) 

From the docs I can only see that the first level allows to specify if that but not a second level relation
https://ravendb.net/docs/article-page/5.4/csharp/document-extensions/counters/counters-and-other-features 

Reason why this is relevant in my case is to avoid a n+1 but for that I need to include the counters otherwise I need to do a roundtrip per doc :) 

Also not sure how this would work with time series either.

Cheers,

**Comment by ayende** [comment]
You can do something like this:

```
declare function includeRelatedCounters(d) {
    var counters = {};
    d.Lines.forEach(l=> {
        include(l.Product); // include related
        counters[l.Product] = counter(l.Product,"⭐⭐⭐⭐⭐")
    });
    return { Doc: d, Counters: counters}
}

from "Orders" as d
select includeRelatedCounters(d)
```

It's not ideal, I agree. Can you explain what it is that you are trying to do?

**Comment by SbiCA** [comment]
@ayende 
How would I do that in c# session.Advanced.RawQuery() or DocumentQuery?

My scenario is the following:

``` mermaid
erDiagram
    User }|..o{ Memberships : has
    User ||..o{ Events : owns
    Memberships }|--|| Events : belongs
```

And I d like to load the data for one user in one roundtrip:
- memberships where userId = $id **using an index** with the fields (userId, eventId)
- include the events (works as of now since it's stored on the membership which user belongs to which event)
- now the missing part I also need all the counters of the `event document` but since I start from the memberships I can't express the include counters here

**Reply by SbiCA** [reply]
I tried the following rql query but somehow without success or are the counters values not visible in the studio response?
``` javascript
declare function includeRelatedCounters(d) {
    var evt = include(d.EventId);
    var numberOfPosts = counter(d.EventId, "NumberOfPosts");
   
 //  "@counters": [
 //         "NumberOfMembers",
 //           "NumberOfPost",
 //          "NumberOfSuggestive",
 //          "NumberOfVideoPost",
 //          "TotalNumberOfImages",
 //           "TotalNumberOfMembers",
 //           "TotalNumberOfVideos"
 //      ],
    return { Doc: evt, Counters: numberOfPosts}
}

from index 'MembersBy/EventAndUser' as d
where UserId = "..."
// include EventId
select includeRelatedCounters(d)
```

**Reply by ayende** [reply]
You cannot *include* the counters of another document, I'm afraid.
You can project them out, however and that should be output for you 

What do you get back?

**Reply by SbiCA** [reply]
@ayende  I created a small repro project https://github.com/SbiCA/raven-repro/tree/main

I got it to work as you can see in this test, what I dislike about this approach is that the information gets decoupled from original document and moved to membership level
[https://github.com/SbiCA/raven-repro/blob/main/Repro/TestIncludeScenario.cs 
[](https://github.com/SbiCA/raven-repro/blob/main/Repro/TestIncludeScenario.cs)](https://github.com/SbiCA/raven-repro/blob/a90f5d573f901c71b7dcdf613276e3d741206d6f/Repro/TestIncludeScenario.cs#L48)

Is there an approach that makes this first class other than use `RawQuery`? 

And could this become a feature that you can specify what you like to inlcude on a related document e.g. ` include(d.EventId, eventIcludeBuilder => eventIcludeBuilder.IncludeCounter(...); eventIcludeBuilder.IncludeTimeseries(...)`? 

I've added the original approach, here you can see the pitfall
https://github.com/SbiCA/raven-repro/blob/a90f5d573f901c71b7dcdf613276e3d741206d6f/Repro/TestIncludeScenario.cs#L27

**Reply by ayende** [reply]
I created an issue about this here: https://issues.hibernatingrhinos.com/issue/RavenDB-20078/Allow-to-include-related-document-counters

For now, I'm afraid that this is not possible except via the javascript option

**Reply by SbiCA** [reply]
OK @ayende  thanks a lot already, I ll try to use that in the meantime. In addition I think you should also consider allowing to include `timeseries` the same way even tough I didn't have that use case yet 😉

**Reply by arekpalinski** [reply]
@SbiCA Is the following syntax of including counters what you're looking for:
https://github.com/ravendb/ravendb/blob/372739ca538938c8306566c55a342304e4d26239/test/SlowTests/Client/Counters/QueryOnCounters.cs#L2336-L2341

?

**Reply by SbiCA** [reply]
Hi @arekpalinski 

Yes indeed 👍  I tried to use that API but somehow the request per session seem still to fail on my end 🥲 https://github.com/SbiCA/raven-repro/blob/d238d0904ebc290aef1384be65c7ebf4fb429328/Repro/TestIncludeScenario.cs#L33

Not sure what I'm doing wrong here?
Cheers

**Reply by SbiCA** [reply]
@arekpalinski  sorry I was missing the include of the event doc.

**Reply by SbiCA** [reply]
So @ayende  I think the issue can be closed 

as you can see here the test is green now 😄 thx @arekpalinski  🎉 
https://github.com/SbiCA/raven-repro/blob/3ddb1630ab9c23173be0d467eb0d6219d6b0639b/Repro/TestIncludeScenario.cs#L27

**Comment by ayende** [comment, marked as answer]
Moving answer from deep inside the thread to here:

```
        var memberships = session.Query<Membership_ByEventAndUserId.Result,Membership_ByEventAndUserId>()
            .Where(m => m.UserId == "User/Ayende")
            .OfType<Membership>()
            .Include(i => i.EventId)
            .Include(i => i.IncludeCounter(m => m.EventId,"Members" ))
            .ToList();
```
